server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,200 @@
+        snapshot.createChildrenMap();
+        // Traverse the tree and build a corresponding tree of MDCSDimensionNode, to pass to MultiDimensionCacheStats.
+        for (DimensionNode child : statsRoot.getChildren().values()) {
+            getCacheStatsHelper(child, snapshot);
+        }
+        return new MultiDimensionCacheStats(snapshot, dimensionNames);
+    }
+
+    private void getCacheStatsHelper(DimensionNode currentNodeInOriginalTree, MDCSDimensionNode parentInNewTree) {
+        MDCSDimensionNode newNode = createMatchingMDCSDimensionNode(currentNodeInOriginalTree);
+        parentInNewTree.getChildren().put(newNode.getDimensionValue(), newNode);
+        if (currentNodeInOriginalTree.hasChildren()) {
+            // not a leaf node
+            for (DimensionNode child : currentNodeInOriginalTree.children.values()) {
+                getCacheStatsHelper(child, newNode);
+            }
+        }
+    }
+
+    private MDCSDimensionNode createMatchingMDCSDimensionNode(DimensionNode node) {
+        CacheStatsCounterSnapshot nodeSnapshot = node.getStats().snapshot();
+        MDCSDimensionNode newNode = new MDCSDimensionNode(node.getDimensionValue(), nodeSnapshot);
+        if (node.getChildren() != null) {
+            newNode.createChildrenMap();
+        }
+        return newNode;
+    }
+
+    public void removeDimensions(List<String> dimensionValues) {
+        assert dimensionValues.size() == dimensionNames.size() : "Must specify a value for every dimension when removing from StatsHolder";
+        CacheStatsCounterSnapshot statsToDecrement = removeDimensionsHelper(dimensionValues, statsRoot, 0);
+        if (statsToDecrement != null) {
+            statsRoot.getStats().subtract(statsToDecrement);
+        }
+    }
+
+    // Returns a CacheStatsCounter object for the stats to decrement if the removal happened, null otherwise.
+    private CacheStatsCounterSnapshot removeDimensionsHelper(List<String> dimensionValues, DimensionNode node, int dimensionValuesIndex) {
+        if (dimensionValuesIndex == dimensionValues.size()) {
+            // Pass up a snapshot of the original stats to avoid issues when the original is decremented by other fn invocations
+            return node.getStats().snapshot();
+        }
+        DimensionNode child = node.getOrCreateChild(dimensionValues.get(dimensionValuesIndex), false);
+        if (child == null) {
+            return null;
+        }
+        CacheStatsCounterSnapshot statsToDecrement = removeDimensionsHelper(dimensionValues, child, dimensionValuesIndex + 1);
+        if (statsToDecrement != null) {
+            // The removal took place, decrement values and remove this node from its parent if it's now empty
+            child.getStats().subtract(statsToDecrement);
+            if (child.getChildren() == null || child.getChildren().isEmpty()) {
+                node.children.remove(child.getDimensionValue());
+            }
+        }
+        return statsToDecrement;
+    }
```

Comment:
I am just wondering if there can be synchronization issues during removal? When all removeDimensionsHelper can be used? When shard is being closed or removed? What if shard is immediately opened or added back to the node, how will we achieve eventually consistency?